### ISSUE TITLE
GCM 

### ISSUE THREAD

**Issue by aficionadoai** [yingkaisha/keras-unet-collection#11]
What does the `axis = -1 ` does ?
```
# Classification-guided Module (CGM)
# ---------------------------------------- #
# dropout --> 1-by-1 conv2d --> global-maxpooling --> sigmoid
X_CGM = X_decoder[-1]
X_CGM = Dropout(rate=0.1)(X_CGM)
X_CGM = Conv2D(filter_num_skip[-1], 1, padding='same')(X_CGM)
X_CGM = GlobalMaxPooling2D()(X_CGM)
X_CGM = Activation('sigmoid')(X_CGM)

CGM_mask = max(X_CGM, axis=-1) # <----- This value could be trained with "none-organ image"

for i in range(len(OUT_stack)):
    if i < len(OUT_stack)-1:
        OUT_stack[i] = multiply([OUT_stack[i], CGM_mask], name='{}_output_sup{}_CGM'.format(name, i))
    else:
        OUT_stack[i] = multiply([OUT_stack[i], CGM_mask], name='{}_output_final_CGM'.format(name))
```

Is this line of code predicting mask or image? Is there way to exclude images that just have background? 
```
CGM_mask = max(X_CGM, axis=-1) # <----- This value could be trained with "none-organ image"
```

If I'm understanding correctly,  the GCM is used to predict if a image blank or not?

**Comment by yingkaisha** [comment]
X_CGM is a two-dimensional tensor of `(batch_size, filter_num_skip[-1])`. The `axis=-1` (or `axis=1`) is just replicating the "argmax" in the Figure 3 of UNET 3+ paper.

**Comment by yingkaisha** [comment]
Yeah, `CGM_mask` is the probability of "if an image is blank or not" and can be trained, but I have not tried. 

Let me know if you have any thoughts.

**Comment by aficionadoai** [comment]
Does this `"if an image is blank or not" and can be trained` you use the output to train another model? 

Or if I want to focus only on non-black images, do I can the -1 to 1 or 0?

**Comment by yingkaisha** [comment]
In that notebook, it was not trained with blank images, and the entire GCM block is updated with the overall loss, not fine-tuned.

Set GCM to one means getting rid of this block. I think it is a good idea, the authors declared benefit by having GCM, but it could be a benefit in their specific case only.

**Comment by aficionadoai** [comment]
@yingkaisha After looking over your example again, I think I understand now. I may have been applying GCM wrong. 

Deep supervision with EfficientNet backbone performed pretty well on the dataset I used.

**Comment by yingkaisha** [comment]
Great to hear that. I will close this issue by the end of the day.
